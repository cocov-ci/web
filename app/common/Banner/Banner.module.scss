@use "sass:color";
@use "sass:map";
@import "styles/variables.module";

@mixin banner-background($color) {
  /* stylelint-disable scss/operator-no-unspaced */
  background-image: radial-gradient(circle at 1.1e+02% 4%, $color 0%, $colorGrayDark 88%);
  /* stylelint-enable scss/operator-no-unspaced */
}

@mixin banner-variation($name) {
  &.#{$name} {
    @include banner-background(map.get($colorBanner, "#{$name}Background"));

    .icon {
      color: map.get($colorBanner, "#{$name}Foreground");
    }
  }
}

.base {
  font-size: 15px;
  display: flex;
  padding: 14px 20px;
  align-items: center;
  color: map.get($colorBanner, "textColor");
  background-color: $colorGrayDark;
  transition: box-shadow 0.2s ease-out 0.1s;

  .icon {
    height: 20px;
    margin-right: 10px;
  }

  .close {
    flex-grow: 1;
    display: flex;
    align-items: center;
    justify-content: end;
    align-content: center;

    button {
      border: none;
      background: none;
      color: map.get($colorBanner, "buttonColor");
      line-height: 0;
      cursor: pointer;
      transition: color 0.2s ease-out;

      &:hover {
        color: color.adjust(map.get($colorBanner, "buttonColor"), $lightness: +50%);
      }
    }
  }

  @include banner-variation("success");
  @include banner-variation("info");
  @include banner-variation("warning");
}

.wrapper {
  &,
  .faces {
    transform-style: preserve-3d;
    transition: transform 0.3s ease-out;
  }

  .faces {
    .base {
      transform: translate(0, -100%) rotateX(90deg);
      transform-origin: bottom;
    }

    &.open {
      transform: translate(0, 50%) rotateX(-90deg);

      .base {
        box-shadow: 7px 2px 40px 0 color.adjust($colorBlack, $alpha: 0.5);
      }
    }
  }
}
