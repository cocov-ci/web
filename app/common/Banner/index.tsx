import classNames from 'classnames'
import { LucideIcon, XIcon } from 'lucide-react'
import React from 'react'

import styles from './Banner.module.scss'

export type BannerVariationProps = 'success' | 'warning' | 'info' | 'neutral'

export type BannerProps = {
  className?: string
  variation?: BannerVariationProps
  children: React.ReactNode
  icon?: LucideIcon
  onClose?: () => void
  open?: boolean
}

const Banner = ({
  className,
  variation = 'neutral',
  icon,
  children,
  onClose,
  open,
}: BannerProps) => {
  const IconComponent = icon

  return (
    <div className={styles.wrapper}>
      <div
        className={classNames(styles.faces, {
          [styles.open]: open,
        })}
      >
        <div className={classNames(styles.base, className)}>
          <div className={styles.shadow} />
          <div className={classNames(styles.innerBase, styles[variation])}>
            <div>
              {IconComponent && (
                <div className={styles.icon}>
                  <IconComponent size={20} />
                </div>
              )}
              <div className={styles.content}>{children}</div>
              <div className={styles.close}>
                <button onClick={onClose}>
                  <XIcon size={20} />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  )
}

export default Banner
