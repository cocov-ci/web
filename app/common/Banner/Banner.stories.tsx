import type { Meta, StoryObj } from '@storybook/react'
import {
  AlertTriangle,
  CheckCircle,
  CheckCircle2,
  Cherry,
  Info,
} from 'lucide-react'

import Banner, { BannerProps } from './index'

const meta: Meta<typeof Banner> = {
  title: 'Common/Banner',
  component: Banner,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof Banner>

const defaultArgs: BannerProps = {
  icon: Info,
  variation: 'info',
  children: (
    <>
      <strong>Notice:</strong> At solmen va esser necessi far uniform
      grammatica, pronunciation e plu commun paroles.
    </>
  ),
}

export const Default: Story = {
  args: { ...defaultArgs },
}

export const Success: Story = {
  args: { ...defaultArgs, variation: 'success', icon: CheckCircle },
}

export const Warning: Story = {
  args: { ...defaultArgs, variation: 'warning', icon: AlertTriangle },
}

export const Neutral: Story = {
  args: { ...defaultArgs, variation: 'neutral', icon: Cherry },
}
