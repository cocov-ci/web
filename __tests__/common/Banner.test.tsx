import { fireEvent, render, screen } from '@testing-library/react'
import { Info } from 'lucide-react'

import Banner from 'app/common/Banner'
import '@testing-library/jest-dom'

describe('common/Banner', () => {
  const { container } = render(
    <Banner open={true} variation="info">
      At solmen va esser necessi far uniform grammatica, pronunciation e plu
      commun paroles.
    </Banner>,
  )

  it('renders Banner', () => {
    expect(
      screen.getByText(
        'At solmen va esser necessi far uniform grammatica, pronunciation e plu commun paroles.',
      ),
    ).toBeVisible()
  })

  it('renders Banner with "icon" param', () => {
    const { container } = render(
      <Banner icon={Info} open={true} variation="info">
        At solmen va esser necessi far uniform grammatica, pronunciation e plu
        commun paroles.
      </Banner>,
    )
    expect(container.querySelectorAll('svg')).toHaveLength(2)
  })

  it('triggers onClose event', () => {
    const handleEvent = jest.fn()

    render(
      <Banner onClose={handleEvent} open={true} variation="info">
        At solmen va esser necessi far uniform grammatica, pronunciation e plu
        commun paroles.
      </Banner>,
    )
    fireEvent.click(screen.getByRole('button'))
    expect(handleEvent).toHaveBeenCalledTimes(1)
  })

  it('renders Banner snapshots', () => {
    expect(container).toMatchSnapshot()
  })
})
