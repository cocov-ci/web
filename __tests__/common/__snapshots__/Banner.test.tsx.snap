// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Banner renders Banner snapshots 1`] = `<div />`;
